File: grimdb/sdk/swift/Sources/GrimlockerSDK/Models/CertificateEntry.swift
import Foundation

public struct CertificateEntry: Codable, Identifiable, Sendable {
    public var id: String
    public var title: String
    public var domain: String
    public var certificate: String
    public var privateKey: String

    enum CodingKeys: String, CodingKey {
        case id, title, domain, certificate
        case privateKey = "private_key"
    }

    public init(id: String = "", title: String, domain: String = "",
                certificate: String = "", privateKey: String = "") {
        self.id = id
        self.title = title
        self.domain = domain
        self.certificate = certificate
        self.privateKey = privateKey
    }

    public var fields: [String: String] {
        [
            "domain": domain,
            "certificate": certificate,
            "private_key": privateKey,
        ]
    }

    public static func from(_ entry: VaultEntry) -> Self {
        .init(
            id: entry.id,
            title: entry.title,
            domain: entry.field("domain"),
            certificate: entry.field("certificate"),
            privateKey: entry.field("private_key")
        )
    }
}

File: grimdb/sdk/csharp/src/Models/CertificateEntry.cs
namespace Grimlocker.Models;

public sealed record CertificateEntry
{
    public string Title { get; init; } = "";
    public string Domain { get; init; } = "";
    public string Certificate { get; init; } = "";
    public string PrivateKey { get; init; } = "";
    public string Id { get; init; } = "";

    public Dictionary<string, string> ToFields() => new()
    {
        ["domain"] = Domain, ["certificate"] = Certificate, ["private_key"] = PrivateKey
    };

    public static CertificateEntry FromEntry(VaultEntry e) => new()
    {
        Id = e.Id, Title = e.Title,
        Domain = e.Fields?.GetValueOrDefault("domain") ?? "",
        Certificate = e.Fields?.GetValueOrDefault("certificate") ?? "",
        PrivateKey = e.Fields?.GetValueOrDefault("private_key") ?? ""
    };
}
